torch/utils/data/dataset.py

```diff
@@ -252,6 +252,41 @@ def __len__(self):
             total += len(d)  # type: ignore
         return total
 
+class ShuffleDataset(IterableDataset[T_co]):
+    r"""Dataset shuffled from the original dataset.
+
+    This class is useful to shuffle the existing dataset. The buffer with buffer_size 
+    is filled with the Dataset first, and then pops out a piece of data randomly for
+    each iteration.
+    For the buffer size smaller than 1, the buffer size is set to 1 representing 
+    non-shuffled dataset. For perfect shuffling, the buffer size is required to be 
+    greater or equal to the size of dataset.
+
+    Arguments:
+        dataset (Dataset): The original Dataset.
+        buffer_size (buffer size): The buffer size for shuffling. 
+    """
+    dataset: IterableDataset[T_co]
+    buf: List[T_co]
+
+    def __init__(self, dataset: IterableDataset[T_co], buffer_size: int) -> None:
+        super(ShuffleDataset, self).__init__()
+        self.dataset = dataset
+        self.buffer_size = buffer_size if buffer_size > 1 else 1
+
+    def __iter__(self) -> Iterator[T_co]:
+        from random import randint 
+        idx = 0
+        self.buf = []
+        for x in self.dataset:
+            if (idx == self.buffer_size):
+                yield self.buf.pop(randint(0, self.buffer_size - 1))
+            else:
+                idx += 1
+            self.buf.append(x)
```

Comment:
I wonder if we introduce a bias in the shuffling by appending to the last element, instead of inserting to the location where the object was last removed. Do you have any insights on this?

Reply:
I think this is fine.

Reply:
When an element comes into buffer as the reservoir, the possibility that any element is popped out should be equal. And, every time the buffer is updated by appending, the different index for each element doesn't destruct the uniform distribution. So, in my opinion, this implementation doesn't ruin the sampling distribution.